src/windows/list.cpp

```diff
@@ -0,0 +1,687 @@
+          result = false;
+          break;
+        }
+        deviceDetailData->cbSize = sizeof(SP_DEVICE_INTERFACE_DETAIL_DATA);
+      } else {
+        device->error = "GetDeviceInterfaceDetailW: "
+          "Couldn't get detail data; Error " + std::to_string(errorCode);
+        result = false;
+        break;
+      }
+    }
+
+    BOOL hasDeviceDetail = SetupDiGetDeviceInterfaceDetailW(
+      hDeviceInfo, &deviceInterfaceData, deviceDetailData, size, &size, NULL);
+
+    if (!hasDeviceDetail) {
+      errorCode = GetLastError();
+      device->error = "Couldn't SetupDiGetDeviceInterfaceDetailW: Error " +
+        std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    printf("[INFO] (%i) SetupDiGetDeviceInterfaceDetailW:\n %s\n",
+      index, WCharToUtf8(deviceDetailData->DevicePath));
+
+    device->device = std::string(WCharToUtf8(deviceDetailData->DevicePath));
+
+    hDevice = CreateFileW(
+      deviceDetailData->DevicePath, handleOpenFlags,
+      FILE_SHARE_READ | FILE_SHARE_WRITE, NULL,
+      OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+
+    if (hDevice == INVALID_HANDLE_VALUE) {
+      errorCode = GetLastError();
+      device->error = "Couldn't open handle to device: Error " +
+      std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    int32_t deviceNumber = GetDeviceNumber(hDevice);
+
+    if (deviceNumber == -1) {
+      device->error = "Couldn't get device number";
+      result = false;
+      break;
+    }
+
+    device->raw = "\\\\.\\PhysicalDrive" + std::to_string(deviceNumber);
+
+    GetMountpoints(deviceNumber, &device->mountpoints);
+
+    hPhysical = CreateFileA(
+      device->raw.c_str(), handleOpenFlags,
+      FILE_SHARE_READ | FILE_SHARE_WRITE, NULL,
+      OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+
+    if (hPhysical == INVALID_HANDLE_VALUE) {
+      errorCode = GetLastError();
+      device->error = "Couldn't open handle to physical device: Error " +
+        std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    if (!GetDeviceSize(hPhysical, device)) {
+      errorCode = GetLastError();
+      device->error = "Couldn't get disk geometry: Error " +
+        std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    if (!GetAdapterInfo(hPhysical, device)) {
+      errorCode = GetLastError();
+      device->error = "Couldn't get device adapter descriptor: Error " +
+        std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    // NOTE: No need to fail over this one,
+    // as we can safely default to a 512B block size
+    if (!GetDeviceBlockSize(hPhysical, device)) {
+      errorCode = GetLastError();
+      printf("[INFO] Couldn't get block size: Error %u\n", errorCode);
+    }
+
+    BOOL isWritable = DeviceIoControl(
+      hPhysical, IOCTL_DISK_IS_WRITABLE, NULL, 0,
+      NULL, 0, &size, NULL);
+
+    device->isReadOnly = !isWritable;
```

Comment:
Just for the sake of consistency, would it also be worth abstracting this out into a `GetDeviceReadOnly(hPhysical, device)` function?